src/analysis/processing/qgsalgorithmroundrastervalues.cpp

```diff
@@ -0,0 +1,263 @@
+  {
+    if ( feedback )
+      feedback->setProgress( 100 * ( ( iterTop / maxHeight * nbBlocksWidth ) + iterLeft / maxWidth ) / nbBlocks );
+    if ( mIsInteger && mDecimalPrecision > -1 )
+    {
+      //nothing to round, just write raster block
+      analysisRasterBlock->setNoDataValue( mInputNoDataValue );
+      destinationRasterProvider->writeBlock( analysisRasterBlock.get(), mBand, iterLeft, iterTop );
+    }
+    else
+    {
+      for ( int row = 0; row < iterRows; row++ )
+      {
+        if ( feedback && feedback->isCanceled() )
+          break;
+        for ( int column = 0; column < iterCols; column++ )
+        {
+          bool isNoData = false;
+          double val = analysisRasterBlock->valueAndNoData( row, column, isNoData );
+          if ( isNoData )
+          {
+            analysisRasterBlock->setValue( row, column, mInputNoDataValue );
+          }
+          else
+          {
+            double roundedVal = mInputNoDataValue;
+            if ( mRoundingDirection == 0 && mDecimalPrecision < 0 )
+            {
+              roundedVal = roundUpBaseN( val, mMultipleOfBaseN );
+            }
+            else if ( mRoundingDirection == 0 && mDecimalPrecision > -1 )
+            {
+              double m = ( val < 0.0 ) ? -1.0 : 1.0;
+              roundedVal = roundUp( val, m, mDecimalPrecision );
+            }
+            else if ( mRoundingDirection == 1 && mDecimalPrecision < 0 )
+            {
+              roundedVal = roundNearestBaseN( val, mMultipleOfBaseN );
+            }
+            else if ( mRoundingDirection == 1 && mDecimalPrecision > -1 )
+            {
+              double m = ( val < 0.0 ) ? -1.0 : 1.0;
+              roundedVal = roundNearest( val, m, mDecimalPrecision );
+            }
+            else if ( mRoundingDirection == 2 && mDecimalPrecision < 0 )
+            {
+              roundedVal = roundDownBaseN( val, mMultipleOfBaseN );
+            }
+            else
+            {
+              double m = ( val < 0.0 ) ? -1.0 : 1.0;
+              roundedVal = roundDown( val,  m, mDecimalPrecision );
+            }
+            //integer values get automatically cast to double when reading and back to int when writing
+            analysisRasterBlock->setValue( row, column, roundedVal );
+          }
+        }
+      }
+      destinationRasterProvider->writeBlock( analysisRasterBlock.get(), mBand, iterLeft, iterTop );
+    }
+  }
+  destinationRasterProvider->setEditable( false );
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), outputFile );
+  return outputs;
+}
+
+double QgsRoundRasterValuesAlgorithm::roundNearest( double &value, double &m, int &decimals )
+{
+  double scaleFactor = std::pow( 10.0, decimals );
+  return ( std::round( value * m * scaleFactor ) / scaleFactor ) * m;
+}
+
+double QgsRoundRasterValuesAlgorithm::roundUp( double &value, double &m, int &decimals )
```

Comment:
These arguments should be passed by value, not reference (by reference is slower then by value for small trivially copied types like int and double)

Reply:
fixed